### PR TITLE
[WIP] DDF-4072 Run solr inside of a docker container during itests

### PR DESCRIPTION
**Pull request by oconnormi**
#### What does this PR do?

This PR introduces some changes to the way itests are run with regards to the external solr process.

With this change solr is now run inside of a docker container instead of manually unpacked and started in the same environment as ddf.

This is the first of a few changes I am proposing for the current itest process which will simply replace the current single solr instance with one that is run inside of a separate container. The next change I will propose, assuming this is merged, will be to have a clean solr instance per test as this should provide a faster clean-up than clearing the single solr instance after each test. This latter change will require more experimentation however and has been left out for now.

Additional Details:

The `distribution/docker/solr` module creates a ddf compatible solr image by building on top of the official solr image and adding in a few things:
* Adds jts-core
* Adds solr-xpath
* Adds a bundled copy of the ddf solr configuration files
* Adds a helper script for creating new ddf solr cores
* Adds pre-created cores for `catalog`, `metacard_cache`, `alerts`, `preferences`, `subscriptions`

#### Who is reviewing it? 

@ahoffer @mcalcote

#### Select relevant component teams:
 
@codice/build @codice/test 

#### Ask 2 committers to review/merge the PR and tag them here.

@rzwiefel @pklinef 

#### How should this be tested?

 1. run a quickbuild
 2. ensure that `docker image ls` shows `codice/ddf-solr 2.14.0-SNAPSHOT` was created
 3. run the itests

#### Any background context you want to provide?

There were several drivers behind this proposed change:
* Inconsistent execution of solr across all environments
* Encountering issues with missing binaries used by solr start scripts
* Encountering issues with certain characters in folder paths automatically used by CI that were breaking solr start scripts
* Occasionally itests would start before solr was ready

#### What are the relevant tickets?
[DDF-4072](https://codice.atlassian.net/browse/DDF-4072)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update any itests that won't work automatically with this change
- [ ] Test on windows

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: distribution/docker/solr/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# DDF Compatible Solr Container
```
Comment (ahoffer) [inline distribution/docker/solr/README.md]: I feel like the definition Solr container should be generic enough (and stable enough) to make it a Codice project outside of DDF.  In fact, I would strongly support a ticket that pulls the creation of the solr-distro.zip file out of DDF and moving it, along with the ability to containerize it, into a new Codice repo.

(Would probably be done as part of another PR)

File: distribution/docker/solr/pom.xml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            <artifactId>solr-xpath</artifactId>
+            <version>${project.version}</version>
+            <exclusions>
+                <exclusion>
+                    <artifactId>*</artifactId>
+                    <groupId>*</groupId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
```
Comment (ahoffer) [inline distribution/docker/solr/pom.xml]: If we make the Solr distro/container a Codice, instead of DDF thing, we should consider moving the schema files and other stuff out. I would recommend publishing them as a single zip artifact because they all get unzipped to the same directory anyway. DDF can just pull that artifact in.

That would clean up this POM file a lot. 

(Would probably be done as part of another PR)

File: distribution/docker/solr/src/main/docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+COPY maven/createCore.sh /usr/local/bin/create-ddf-core
+RUN chmod 755 /usr/local/bin/create-ddf-core
+RUN chown -R solr:solr /seed
+USER solr
+
+# Set ENV vars for use by create-ddf-core script
+ENV CONFIG_SRC=/seed
+ENV CORE_DEST=/opt/solr/server/solr/cores
+RUN mkdir -p $CORE_DEST
+RUN /usr/local/bin/create-ddf-core catalog metacard_cache alerts preferences subscriptions
```
Comment (ahoffer) [inline distribution/docker/solr/src/main/docker/Dockerfile]: This is pretty cool.
Downstream projects use additional cores not listed in their tests. Is there a way we can let those projects add more cores without editing any of these files?

**Comment by ahoffer** [comment]
I feel like there might be an itest out there that dynamically create a core. You'll know if the itest fails with an error message about not being able to copy files to the solr data dir.

File: distribution/docker/solr/README.md (hunk 1 above)
Comment (oconnormi) [reply distribution/docker/solr/README.md]: Yeah I was actually talking about exactly that with @lessarderic earlier since it seems like that would be the best approach. I think @pklinef had some concerns about that however since we would (in order to make dependencies flow only in a single direction) need to move the solr schemas out and probably any code that is affected by changes to those schemas

File: distribution/docker/solr/src/main/docker/Dockerfile (hunk 3 above)
Comment (oconnormi) [reply distribution/docker/solr/src/main/docker/Dockerfile]: With this current approach they could package a distro specific docker image by building on top of the one produced here and simply having a `RUN create-ddf-core <cores>` to add the additional cores

I'm already thinking of moving away from that method however and just having an way of creating them dynamically when starting a new solr container. Possibly by passing in a list of core names as an env var. This would the single image more usable by downstreams without having to build a new image on top

File: distribution/docker/solr/pom.xml (hunk 2 above)
Comment (oconnormi) [reply distribution/docker/solr/pom.xml]: Yep agree completely. I was already thinking of doing that once I realized I needed all these exclusions

File: distribution/docker/solr/README.md (hunk 1 above)
Comment (oconnormi) [reply distribution/docker/solr/README.md]: The other thing I'm not sure on is how we would handle versioning of the solr-distro stuff. Would we have the separate repo just use the same version number as the target ddf version (probably simplest, but feels slightly odd) or would we have it use the solr version that it is building on top of?

Another option could be a combo of the two where it would produce a version like <target_ddf_version>-<solr-version>.

Additionally it could simply start off with a completely new version number if all the solr related code lived in the new repo. With this approach consuming projects would manage updating the version of ddf-solr (for lack of a better name at this time) as needed.

**Comment by rzwiefel** [comment]
@oconnormi Whats the status on this PR? Is it still a WIP?

**Comment by oconnormi** [comment]
Gonna close this since some of these changes will be incorporated into a new pr
